src/Type/Doctrine/QueryBuilder/QueryBuilderGetQueryDynamicReturnTypeExtension.php

```diff
@@ -32,6 +32,21 @@
 class QueryBuilderGetQueryDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
 {
 
+	/**
+	 * Those are critical methods where we need to understand arguments passed to them, the rest is allowed to be more dynamic
+	 * - this list reflects what is implemented in QueryResultTypeWalker
+	 */
+	private const METHODS_AFFECTING_RESULT_TYPE = [
+		'add',
```

Comment:
Somehow we must not forget to update this list when the type inference gets smarter and takes `andWhere` etc. into account.